basil/HL/JtagMaster.py

```diff
@@ -0,0 +1,287 @@
+        return self.MEM_BYTES
+
+    def set_data(self, data, addr=0):
+        """
+        Sets data for outgoing stream
+        """
+        if self._mem_bytes < len(data):
+            raise ValueError(
+                "Size of data (%d bytes) is too big for memory (%d bytes)"
+                % (len(data), self._mem_bytes)
+            )
+        self._intf.write(self._conf["base_addr"] + self._mem_offset + addr, data)
+
+    # This needs to be changed to return written value
+    def get_data(self, size=None, addr=None):
+        """
+        Gets data for incoming stream
+        """
+        # readback memory offset
+        if addr is None:
+            addr = self._mem_bytes
+
+        if size and self._mem_bytes < size:
+            raise ValueError("Size is too big")
+
+        if size is None:
+            return self._intf.read(
+                self._conf["base_addr"] + self._mem_offset + addr, self._mem_bytes
+            )
+        else:
+            return self._intf.read(
+                self._conf["base_addr"] + self._mem_offset + addr, size
+            )
+
+    def scan_ir(self, data):
+        """
+        Data must be a list of BitLogic
+        """
+        if type(data[0]) == BitLogic:
+            pass
+        else:
+            raise TypeError(
+                "Type of data not supported: got",
+                type(data[0]),
+                " and support only Bitlogic",
+            )
+
+        # Set value to pass all data
+        bit_number = sum(len(x) for x in data)  # calculate number of bit to transmit
+        if bit_number < self._mem_bytes * 8:
+            self.set_size(bit_number)
+        else:
+            raise ValueError(
+                "Size is too big for memory: got %d and memory is: %d" % bit_number,
+                self._mem_bytes * 8,
+            )
+
+        data_byte = self._bitlogic2bytes(data)
+        self.set_data(data_byte[::-1])
+
+        self.set_word_count(1)
+        self.set_operation(0)
+
+        self.start()
+        while not self.is_ready:
+            pass
+
+        received_data = self.get_data(size=len(data_byte))
+        ret = self._bytes2bitlogic(received_data, bit_number)
+        rlist = self._split_bitlogic(ret, data)
+
+        return rlist
+
+    def scan_dr(self, data, words=1):
+        """
+        Data must be a list of BitLogic or string of raw data
+        """
+        if type(data[0]) == BitLogic or type(data[0]) == str:
+            pass
+        else:
+            raise TypeError(
+                "Type of data not supported: got",
+                type(data[0]),
+                " and support only str and Bitlogic",
+            )
+
+        bit_number = sum(len(x) for x in data)
+        if bit_number < self._mem_bytes * 8:
+            self.set_size(bit_number)
+        else:
+            raise ValueError(
+                "Size is too big for memory: got %d and memory is: %d" % bit_number,
+                self._mem_bytes * 8,
+            )
```

Comment:
Looks like repeating code like in `scan_ir`? Maybe function?

Reply:
yes, no problems